homeassistant/components/intellifire/config_flow.py

```diff
@@ -4,60 +4,138 @@
+        LOGGER.debug("Not Configured Hosts: %s", str(self._not_configured_hosts))
+
+    async def async_step_local_config(self, user_input=None):
+        """Handle local ip configuration."""
+
         errors = {}
+        placeholder = {}
+        local_schema = vol.Schema({vol.Required(CONF_HOST): str})
+
+        if user_input is None:
+
+            if self._not_configured_hosts:
+                user_input = {CONF_HOST: self._not_configured_hosts[0]}
+                local_schema = vol.Schema(
+                    {vol.Required(CONF_HOST, default=user_input[CONF_HOST]): str}
+                )
+            else:
+                if self._discovered_hosts:
+                    # If there were discoverd hosts AND all got filtered out fire an error message
+                    errors["base"] = "already_discovered"
+                    LOGGER.debug("All discovered fireplaces have been configured")
 
-        try:
-            serial = await validate_input(self.hass, user_input[CONF_HOST])
-        except (ConnectionError, ClientConnectionError):
-            errors["base"] = "cannot_connect"
         else:
-            await self.async_set_unique_id(serial)
+            # There was User Input
+            placeholder = {CONF_HOST: user_input[CONF_HOST]}
+
             self._abort_if_unique_id_configured(
-                updates={CONF_HOST: user_input[CONF_HOST]}
+                updates={
+                    CONF_HOST: user_input[CONF_HOST],
+                }
             )
 
-            return self.async_create_entry(
-                title="Fireplace",
-                data={CONF_HOST: user_input[CONF_HOST]},
-            )
+            try:
+                # Validate the ip address
+                serial = await validate_host_input(user_input[CONF_HOST])
+            except (ConnectionError, ClientConnectionError):
+                errors["base"] = "cannot_connect"
+                local_schema = vol.Schema(
+                    {vol.Required(CONF_HOST, default=user_input[CONF_HOST]): str}
+                )
+            else:
+                await self.async_set_unique_id(serial)
+
+                return self.async_create_entry(
+                    title="Fireplace",
+                    data={
+                        CONF_HOST: user_input[CONF_HOST],
+                    },
+                )
+
+        return self.async_show_form(
+            step_id="local_config",
+            errors=errors,
+            description_placeholders=placeholder,
+            data_schema=local_schema,
+        )
+
+    async def async_step_pick_device(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Pick which device to configure."""
+
+        if user_input is not None:
+            # Update not_configured_hosts in order to trigger correct
+            # behavior in next step
+            self._not_configured_hosts = [user_input[CONF_HOST]]
```

Comment:
This is still effectively passing user input between steps.

Reply:
Could you elaborate why this is actually an issue - and here is what my flow was supposed to be doing.

As I understand the forms they basically display and then re-flow into themselves with `user_input` being set.  

So on the picker 

<img width="272" alt="image" src="https://user-images.githubusercontent.com/6491743/158020070-da7bdd0a-474f-4998-a3c6-f117708f0d66.png">

it will display options to the user. They will select on and then it will "re_flow" into itself with the selected data.

In this example it would reflow into its self and `user_input` will be `192.168.1.73`.


From there we flow into `local_config` and then trigger the validation logic -> which only happens if `user_input is not None` and start running the validation logic:

When validation fails it will set `errors` and `placeholder` and return this error message

<img width="514" alt="image" src="https://user-images.githubusercontent.com/6491743/158020431-30b76e45-ac3c-4b2f-8472-eb1736738550.png">

It uses the same validation as if the user had gone directly into manual entry mode.

What I'm trying to do - and I'm open to other suggestions is basically this:

**Allow user to pick a fireplace** -> **Validate** -> (if bad) -> **Display an error version of the manual picking interface**

I'm going to wait on your response before I make any more changes ... 

(Although I did restore the `user_input` passing because thats how I trigged this logic)

_also the kids just got up so it will probably be a bit until I'm back to this._

Reply:
The flow system is designed so you present a form, consume the input in that step, and display errors in the step the error occurred so the user can correct it and try again or handle the problem.   It is confusing for the user to be unexpectedly moved between steps since when something goes wrong they end up in an error state in another screen.

Additionally, we want to handle the branching for the user input in the step it is collected.  By handling it in another step makes everything much more difficult to follow and it leads to discussions like https://github.com/home-assistant/core/pull/67002#discussion_r825300663 where everyone ends up frustrated since the code path is complex to follow.

We should take user input, process the input and then move the user to the next step.

For example `async_step_pick_device` shouldn't pass the user input it collected  to `async_step_local_config`.  In the current scenario if the user chooses the wrong fireplace and they can't connect to it, they don't get back to pick the other one as they have been moved another step and lost context for the error.

```python3

    async def async_step_pick_device(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Pick which device to configure."""

        if user_input is not None:
            # Update not_configured_hosts in order to trigger correct
            # behavior in next step

            return await self.async_step_local_config(user_input=user_input)

        return self.async_show_form(
            step_id="pick_device",
            data_schema=vol.Schema(
                {vol.Required(CONF_HOST): vol.In(self._not_configured_hosts)}
            ),
        )
```

You could rewrite something like this:

```python3

    async def async_step_pick_device(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Pick which device to configure."""

        if user_input is not None:
            host = user_input[CONF_HOST)
            if not (errors := validate_function(user_input[CONF_HOST))):
                return self._async_create_entry_for_host(host)
   
        return self.async_show_form(
            step_id="pick_device",
            errors=errors,
            data_schema=vol.Schema(
                {vol.Required(CONF_HOST): vol.In(self._not_configured_hosts)}
            ),
        )
```

If you send them here before asking if they want to enter an ip address you can add an option to select an ip address manually, and then send them to a manual entry step if they select it by calling something similar `return await self.async_step_manual_entry()`without passing `user_input`

Reply:
I like this idea ... I'll see if I have time to rewrite it today...